ext2/ext2.c

```diff
@@ -1551,7 +1558,9 @@ static int _count_dirents(
     if ((end - p) % ext2->block_size)
         ERAISE(-EINVAL);
 
-    while (p < end)
+    /* The end of the block might not have a full ext2_dirent structure. Make
+     * sure the fixed-length header portion is in range before accessing */
+    while (p + sizeof(ext2_dirent_t) - EXT2_PATH_MAX < end)
```

Comment:
The `if (p !=end)` check after the while loop might fail if its the last block with trailing zeros?

Reply:
changed the logic to `if (p > end)` in the latest update